apps/frontend/src/components/ChessBoard.tsx

```diff
@@ -80,6 +80,24 @@ export const ChessBoard = ({
   const isMyTurn = myColor === chess.turn();
   const [legalMoves, setLegalMoves] = useState<string[]>([]);
 
+  // Get the square king is in
+  let kingSquare: string = '';
+  for (let i = 0; i < 8; i++) {
```

Comment:
Can u move this fn to a saparate file?

Reply:
Need a better logic to find it rather than looping all through

Reply:
> Need a better logic to find it rather than looping all through

I don't think there is a better way to find the present square of king. The chessjs library also doesn't provide anything like that

Reply:
> > Need a better logic to find it rather than looping all through
> 
> 
> 
> I don't think there is a better way to find the present square of king. The chessjs library also doesn't provide anything like that

No, there should be a better algo. n^2 is a worst thing to do in every move. Try traversing from the king's location to all the possible places.

Reply:
```javascript
// Define a function to check if a given king position is under attack by opponent pieces
function isUnderAttack(kingPos, board) {
    // Define the directions of possible attacks for each type of piece
    const directions = {
        'P': [[1, 1], [1, -1]],
        'B': [[1, 1], [1, -1], [-1, 1], [-1, -1]],
        'R': [[1, 0], [-1, 0], [0, 1], [0, -1]],
        'Q': [[1, 0], [-1, 0], [0, 1], [0, -1], [1, 1], [1, -1], [-1, 1], [-1, -1]],
        'K': [[1, 1], [1, -1], [-1, 1], [-1, -1], [1, 0], [-1, 0], [0, 1], [0, -1]],
        'N': [[1, 2], [1, -2], [-1, 2], [-1, -2], [2, 1], [2, -1], [-2, 1], [-2, -1]]
    };

    for (const piece in directions) {
        for (const direction of directions[piece]) {
            let [dx, dy] = direction;
            let x = kingPos[0] + dx;
            let y = kingPos[1] + dy;

            if (piece === 'N' && (x >= 0 && x < 8 && y >= 0 && y < 8)) {
                if (board[x][y][0] !== board[kingPos[0]][kingPos[1]][0] && board[x][y][1] === 'N') {
                    return true;
                }
                continue;
            }

            while (x >= 0 && x < 8 && y >= 0 && y < 8) {
                if (board[x][y] !== '--') {
                    if (board[x][y][0] !== board[kingPos[0]][kingPos[1]][0]) {
                        if (piece === 'P' || piece === 'Q') {
                            return true;
                        } else if (piece === 'B' && (dx === 1 || dx === -1)) {
                            return true;
                        } else if (piece === 'R' && (dx === 0 || dy === 0)) {
                            return true;
                        }
                    }
                    break;
                }
                x += dx;
                y += dy;
            }
        }
    }
    return false;
}


// Function to check if both kings are in check
function bothKingsInCheck(whiteKingPos, blackKingPos, board) {
    const whiteKingCheck = isUnderAttack(whiteKingPos, board);
    const blackKingCheck = isUnderAttack(blackKingPos, board);
    return {whiteKingCheck, blackKingCheck};
}
```

So, this can be a more optimal approach. Rather than looping through the board with a double for loop.